### PR TITLE
[SPARK-13484][SQL] Prevent illegal NULL propagation when filtering outer-join results

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?

This pr is to prevent illegal NULL propagation in the query below;

```
val a = sqlContext.range(10).select(col("id"), lit(0).as("count"))
val b = sqlContext.range(10).select((col("id") % 3).as("id")).groupBy("id").count()
a.join(b, a("id") === b("id"), "left_outer").filter(b("count").isNull)
```

It returns nothing because `b("count")` is not nullable and the filter condition is always false by `Optimizer`.
## How was this patch tested?

Added a test for the query above in `DataFrameJoinSuite`.

### CODE REVIEW

**Comment by maropu** [comment]
Jenkins, retest this please.

**Comment by mengxr** [comment]
cc @yhuai

**Comment by yhuai** [comment]
@maropu Thank you for the PR. My thought is that we may need to have a place to correct those nullable fields in the analyzer. Let me also think about it.

**Comment by maropu** [comment]
@yhuai okay.

**Comment by maropu** [comment]
@yhuai I added a new role 'SolveIllegalReference` to solve these kinds of illegal references.

**Comment by maropu** [comment]
Jenkins, retest this please.

**Comment by maropu** [comment]
Jenkins, retest this please.

**Comment by maropu** [comment]
@yhuai ping

**Comment by rxin** [comment]
cc @cloud-fan

**Comment by cloud-fan** [comment]
I think the fundamental problem is, we give users the resolved attribute but it may not be the real column when using it. For example, `b("count")` actually is not the real column of the join. Instead of adding some special handling, how about my proposal at https://github.com/apache/spark/pull/11632?

**Comment by maropu** [comment]
@cloud-fan If your pull request (#11632) merged, I think the query in the top throws analysis exception, right? `SPARK-13484` essentially indicates that the kinds of queries should be correctly resolved in terms of user's usability. Anyway, I agree with your idea in #11632, so I'd like to discuss this based on #11632.
What do you think? cc: @mengxr

**Comment by maropu** [comment]
Jenkins, retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -1443,6 +1445,32 @@ class Analyzer(
...
+
+    private def replaceReferences(e: Expression, attrMap: AttributeMap[Attribute]) = e.transform {
+      case a: AttributeReference => attrMap.get(a).getOrElse(a)
+    }
+
+    def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+      case q: LogicalPlan =>
+        q.transform {
+          case f @ Filter(filterCondition, ExtractJoinOutputAttributes(join, joinOutputMap)) =>
+            f.copy(condition = replaceReferences(filterCondition, joinOutputMap))
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: How about we use a `q.transformUp` to fix the nullability in a bottom-up way? For every node, we create an `AttributeMap` using the output of its child. Then, we use `transformExpressions` to fix the nullability if necessary.  Let me try it out and ping you when I have a version.

Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: okay, I wait your ping.

Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: https://github.com/apache/spark/pull/13290/files

Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: okay, I'll check it.

**Pull request closed without merging**